### PR TITLE
Structurize and accelerate sending metrics

### PR DESCRIPTION
**Pull request by Buristan**

### CODE REVIEW

File: app/socket.lua [hunk 1]
```diff
@@ -92,19 +84,28 @@ function M:open()
...
-	close(self.sockfd)
+	C.close(self.sockfd)
 	ngx.log(ngx.WARN, string.format("Socket on %s:%d closed", self.host, tonumber(self.port)))
 end
 
-function M:send(msg)
-	msg = tostring(msg)
-	local r = sendto(self.sockfd, msg, #msg, 0, self.dest_addr, self.addr_len)
+function M:send(msg, len)
+	if type(msg) == 'cdata' then
```
Comment (Victory12) [inline app/socket.lua:92]: а в каких случаях ты хочешь посылать cdata в не просто строку?

File: app/socket.lua [hunk 2]
```diff
@@ -49,19 +49,11 @@ ffi.cdef[[
...
-local close      = C.close
-local htons      = C.htons
-local sendto     = C.sendto
-local strerror   = C.strerror
-
 function M:new(params)
 	local sock = {
 		host = params.host;
 		port = params.port;
+		sockfd = -1;
```
Comment (Victory12) [inline app/socket.lua:56]: а зачем? оно же перезатрется

File: app/graphite.lua [hunk 3]
```diff
@@ -1,95 +1,92 @@
...
 end
 
 function M:send()
-	local success, operation = pcall(self.metric.operation_qualifier)
+	-- operation[1] - metric name
+	-- operation[2] - list of metric's types
+	-- operation[3] - reserved for internal usage
+	-- operation[4] - reserved for internal usage
+	local success, operation = pcall(self.qualifier, self)
 	if not success then return ngx.log(ngx.ERR, operation) end
```
Comment (Victory12) [inline app/graphite.lua]: ngx.log(ngx.ERR, operation) если operation это таблица то как оно будет логироваться?

File: app/graphite.lua [hunk 4]
```diff
@@ -1,95 +1,92 @@
...
 		error(result)
 	end
-
+	
 	graph.sock = result
-	setmetatable(graph , self)
+	setmetatable(graph, self)
 	self.__index = self
 	
-	ngx.log(ngx.WARN, "Graphite created")
```
Comment (Victory12) [inline app/graphite.lua:82]: а может вернем лог? чем он помешал

File: app/graphite.lua [hunk 5]
```diff
@@ -1,95 +1,92 @@
...
 
 function M:send()
-	local success, operation = pcall(self.metric.operation_qualifier)
+	-- operation[1] - metric name
+	-- operation[2] - list of metric's types
+	-- operation[3] - reserved for internal usage
+	-- operation[4] - reserved for internal usage
+	local success, operation = pcall(self.qualifier, self)
 	if not success then return ngx.log(ngx.ERR, operation) end
+	if not operation then return end
```
Comment (Victory12) [inline app/graphite.lua:81]: тут тоже бы лог а не просто ретерн

File: app/graphite.lua [hunk 6]
```diff
@@ -1,95 +1,92 @@
...
+	if not operation[2] then
```
Comment (Victory12) [inline app/graphite.lua:82]: мб проверить что type(operation) == 'table'

File: app/graphite.lua [hunk 7]
```diff
@@ -1,95 +1,92 @@
...
+ffi.cdef[[
+	int snprintf(char *str, size_t, const char *format, ...);
+]]
+
+local C = ffi.C
+-- max uri length is 16 Kbyte on nginx
+local buff_size = 16 * 1024
+local message = ffi.new("char[?]", buff_size)
+
+function M:transmit(pref, t, op, field, value)
```
Comment (Victory12) [inline app/graphite.lua]: давай заменим t---> type / op --> operation
понятнее же будет

File: app/socket.lua [hunk 8]
```diff
@@ -72,18 +64,18 @@ end
...
-			error "Impossible to open socket"
+			error("Impossible to open socket", 2)
 		end
 	end
 	
 	local addr = ffi.new("struct sockaddr_in")
-	addr.sin_family = AF_INET
-	addr.sin_port = htons(self.port)
-	inet_aton(self.host, addr.sin_addr)
+	addr.sin_family = C.AF_INET
```
Comment (okoshovetc) [inline app/socket.lua]: test comment2

File: app/graphite.lua (hunk 5 above)
Comment (Buristan) [reply app/graphite.lua:81]: This is default behavior when we don't want to send any stat by method we can return nothing
Don't want to do this slower

File: app/graphite.lua (hunk 6 above)
Comment (Buristan) [reply app/graphite.lua:82]: No, if it isn't table, this is your fault

File: app/socket.lua (hunk 1 above)
Comment (Buristan) [reply app/socket.lua:92]: Always, but for flexibility can keep variant with lua string

File: app/socket.lua (hunk 2 above)
Comment (Buristan) [reply app/socket.lua:56]: It was better if somebody try to use close or send function with fd with value -1instead nil

File: app/graphite.lua (hunk 4 above)
Comment (Buristan) [reply app/graphite.lua:82]: It is uninformative at all. The most important logs (socket logs) include in themselves info about host:port on which they was open

File: app/socket.lua (hunk 1 above)
Comment (Victory12) [reply app/socket.lua:92]: can you give an example of usage "send" function?

File: app/graphite.lua (hunk 4 above)
Comment (Victory12) [reply app/graphite.lua:82]: ok

File: app/socket.lua (hunk 2 above)
Comment (Victory12) [reply app/socket.lua:56]: ok

File: app/graphite.lua (hunk 5 above)
Comment (Victory12) [reply app/graphite.lua:81]: by default qualifier function return ```{method, {hits, ...}}```
so why do you think that no operation is default behavior ?

File: app/graphite.lua (hunk 6 above)
Comment (Victory12) [reply app/graphite.lua:82]: readable error message can help to find mistake

File: app/graphite.lua (hunk 3 above)
Comment (Victory12) [reply app/graphite.lua]: up, no answer for this question

File: app/socket.lua (hunk 1 above)
Comment (Buristan) [reply app/socket.lua:92]: send from graphite:transmit
https://github.com/moonlibs/nginx-graphite/pull/5/files#diff-83f05b2b6780b4c2220983b7460079bfR20

File: app/graphite.lua (hunk 5 above)
Comment (Buristan) [reply app/graphite.lua:81]: Default behavior of this method, not qualifier. If somebody doesn't want to send any metrics, he can return empty operation.

File: app/graphite.lua (hunk 6 above)
Comment (Buristan) [reply app/graphite.lua:82]: It is easy to understand what went wrong if you read the comment here: 
https://github.com/moonlibs/nginx-graphite/pull/5/files#diff-83f05b2b6780b4c2220983b7460079bfR76

**Pull request merged**